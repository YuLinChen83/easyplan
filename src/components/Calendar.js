import React, { Component } from 'react';
import dateFns from 'date-fns';
import './Calendar.css';

export const CalendarHeader = ({ currentMonth, prevMonth, nextMonth }) => {
  const dateFormat = 'MMMM YYYY';
  return (
    <div className="header row flex-middle">
      <div className="col col-start">
        <div
          id="prevMonth"
          role="presentation"
          className="icon"
          onClick={prevMonth}
        >
          chevron_left
        </div>
      </div>
      <div className="col col-center">
        <span>{dateFns.format(currentMonth, dateFormat)}</span>
      </div>
      <div
        id="nextMonth"
        role="presentation"
        className="col col-end"
        onClick={nextMonth}
      >
        <div className="icon">chevron_right</div>
      </div>
    </div>
  );
};
const CalendarDays = ({ currentMonth }) => {
  const dateFormat = 'dddd';
  const days = [];
  const startDate = dateFns.startOfWeek(currentMonth);

  for (let i = 0; i < 7; i += 1) {
    days.push(
      <div className="col col-center" key={i}>
        {dateFns.format(dateFns.addDays(startDate, i), dateFormat)}
      </div>,
    );
  }

  return <div className="days row">{days}</div>;
};
export const CalendarCells = ({
  currentMonth,
  selectedDate,
  onDateClick,
  preferDate,
  unavailableDate,
}) => {
  const monthStart = dateFns.startOfMonth(currentMonth);
  const monthEnd = dateFns.endOfMonth(monthStart);
  const startDate = dateFns.startOfWeek(monthStart);
  const endDate = dateFns.endOfWeek(monthEnd);

  const dateFormat = 'D';
  const rows = [];

  let days = [];
  let day = startDate;
  let formattedDate = '';
  let preferCount = 0;

  while (day <= endDate) {
    for (let i = 0; i < 7; i += 1) {
      formattedDate = dateFns.format(day, dateFormat);
      const cloneDay = day;
      const compositeDate = `${day.getFullYear()}-${(day.getMonth() + 1 < 10
        ? '0'
        : '')
        + (day.getMonth() + 1)}-${(day.getDate() < 10 ? '0' : '')
        + day.getDate()}`;

      preferCount = preferDate[compositeDate];

      days.push(
        <div
          role="presentation"
          className={`dateDiv col cell ${
            // eslint-disable-next-line no-nested-ternary
            !dateFns.isSameMonth(day, monthStart)
              ? 'disabled'
              : dateFns.isSameDay(day, selectedDate)
                ? 'selected'
                : ''
          } 
        ${compositeDate in preferDate ? 'prefer' : ''}
          `}
          key={day}
          onClick={() => onDateClick(dateFns.parse(cloneDay))}
        >
          <div className="statistics">
            {preferCount ? `可以人數${preferCount}` : ''}

            {unavailableDate.indexOf(compositeDate) !== -1 ? (
              <span className="unavailableMark">（有人不行）</span>
            ) : (
              ''
            )}
          </div>
          <span className="number">{formattedDate}</span>
          <span className="bg">{formattedDate}</span>
        </div>,
      );
      day = dateFns.addDays(day, 1);
    }
    rows.push(
      <div className="row" key={day}>
        {days}
      </div>,
    );
    days = [];
  }
  return <div className="body">{rows}</div>;
};

export const NameFilter = ({ userList, filterByName, setFilterName }) => (
  <div>
    <select
      id="nameFilter"
      value={filterByName}
      onChange={e => setFilterName(e.target.value)}
    >
      <option value="ALL">顯示全部</option>
      {userList.map(name => (
        <option key={name} value={name}>
          {name}
        </option>
      ))}
    </select>
  </div>
);

class Calendar extends Component {
  constructor() {
    super();
    const mockDate = new Date();
    this.state = {
      currentMonth: mockDate,
      selectedDate: mockDate,
    };
  }

  onDateClick = (day) => {
    this.setState({
      selectedDate: day,
    });
  };

  nextMonth = () => {
    const { currentMonth } = this.state;
    this.setState({
      currentMonth: dateFns.addMonths(currentMonth, 1),
    });
  };

  prevMonth = () => {
    const { currentMonth } = this.state;
    this.setState({
      currentMonth: dateFns.subMonths(currentMonth, 1),
    });
  };

  render() {
    const {
      dateStatistics,
      userList,
      filterByName,
      setFilterName,
    } = this.props;
    const { preferDate, unavailableDate } = dateStatistics;
    const { currentMonth, selectedDate } = this.state;
    return (
      <div>
        <NameFilter
          userList={userList}
          filterByName={filterByName}
          setFilterName={setFilterName}
        />
        <div className="calendar">
          <CalendarHeader
            currentMonth={currentMonth}
            prevMonth={this.prevMonth}
            nextMonth={this.nextMonth}
          />
          <CalendarDays currentMonth={currentMonth} />
          <CalendarCells
            currentMonth={currentMonth}
            selectedDate={selectedDate}
            onDateClick={this.onDateClick}
            preferDate={preferDate}
            unavailableDate={unavailableDate}
          />
        </div>
      </div>
    );
  }
}

export default Calendar;
