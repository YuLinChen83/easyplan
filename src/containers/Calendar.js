import { connect } from 'react-redux';
import { setFilterName, updateByDate } from '../actions';
import Calendar from '../components/Calendar';

const dateStatistics = (filter, selectedList) => {
  const filterselectedList = filter !== 'ALL'
    ? selectedList.filter(item => item.name === filter)
    : selectedList;
  const { preferDate, unavailableDate } = filterselectedList.reduce(
    (parentPrev, parentCur) => ({
      preferDate: [...parentPrev.preferDate, ...parentCur.preferDate],
      unavailableDate: [
        ...new Set([
          ...parentPrev.unavailableDate,
          ...parentCur.unavailableDate,
        ]),
      ],
    }),
    { preferDate: [], unavailableDate: [] },
  );
  const preferDateStatistics = preferDate.reduce((prev2, cur2) => {
    // eslint-disable-next-line no-param-reassign
    prev2[cur2] = (prev2[cur2] || 0) + 1;
    // eslint-enable
    return prev2;
  }, {});
  const finalDateStatistics = {
    preferDate: preferDateStatistics,
    unavailableDate,
  };
  return finalDateStatistics;
};
const mapStateToProps = state => ({
  dateStatistics: dateStatistics(state.filterByName, state.selectedList),
  userList: state.userList,
  filterByName: state.filterByName,
});

const mapDispatchToProps = dispatch => ({
  setFilterName: filter => dispatch(setFilterName(filter)),
  updateByDate: (dateType, selectedDate) => dispatch(updateByDate(dateType, selectedDate)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Calendar);
