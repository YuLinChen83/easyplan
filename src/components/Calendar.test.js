import React from 'react';
import { shallow } from 'enzyme';
import Calendar, {
  NameFilter,
  CalendarHeader,
  CalendarCells,
} from './Calendar';

const setup = (props) => {
  const enzymeWrapper = shallow(<Calendar {...props} />);
  return {
    props,
    enzymeWrapper,
  };
};
const setupNameFilter = (props) => {
  const enzymeWrapper = shallow(<NameFilter {...props} />);
  return {
    props,
    enzymeWrapper,
  };
};
const setupCalendarHeader = (props) => {
  const enzymeWrapper = shallow(<CalendarHeader {...props} />);
  return {
    props,
    enzymeWrapper,
  };
};
const setupCalendarCells = (props) => {
  const enzymeWrapper = shallow(<CalendarCells {...props} />);
  return {
    props,
    enzymeWrapper,
  };
};

describe('<Calendar/ >', () => {
  const initProps = {
    dateStatistics: { preferDate: {}, unavailableDate: [] },
    userList: [],
    filterByName: 'ALL',
    preferDate: {},
    unavailableDate: [],
    setFilterName: jest.fn(),
    prevMonth: jest.fn(),
    nextMonth: jest.fn(),
    onDateClick: jest.fn(),
    currentMonth: 2,
    selectedDate: 2,
  };

  describe('render', () => {
    it('no crashing', () => {
      const { enzymeWrapper } = setup(initProps);
      expect(enzymeWrapper.length).toBe(1);
      expect(enzymeWrapper).toMatchSnapshot();
    });

    it('should render self and subcomponents', () => {
      const { enzymeWrapper } = setup(initProps);
      expect(enzymeWrapper.find('NameFilter').length).toEqual(1);
      expect(enzymeWrapper.find('CalendarHeader').length).toEqual(1);
      expect(enzymeWrapper.find('CalendarDays').length).toEqual(1);
      expect(enzymeWrapper.find('CalendarCells').length).toEqual(1);
    });

    it('should trigger "onDateClick"', () => {
      const { enzymeWrapper } = setupCalendarCells(initProps);
      enzymeWrapper
        .find('.dateDiv')
        .first()
        .simulate('click');
      expect(initProps.onDateClick).toHaveBeenCalled();
    });

    it('should trigger "prevMonth", "nextMonth"', () => {
      const { enzymeWrapper } = setupCalendarHeader(initProps);
      enzymeWrapper.find('#prevMonth').simulate('click');
      expect(initProps.prevMonth).toHaveBeenCalled();
      expect(enzymeWrapper.props().currentMonth).toBe(1);

      enzymeWrapper.find('#nextMonth').simulate('click');
      expect(initProps.nextMonth).toHaveBeenCalled();
      expect(enzymeWrapper.props().currentMonth).toBe(3);
    });

    it('should trigger "setFilterName"', () => {
      const { enzymeWrapper } = setupNameFilter(initProps);
      const event = { target: { value: 'Jack' } };
      enzymeWrapper.find('#nameFilter').simulate('change', event);
      expect(initProps.setFilterName).toHaveBeenCalledWith(event.target.value);
    });
  });
});
