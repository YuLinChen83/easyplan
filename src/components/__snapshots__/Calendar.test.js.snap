// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar/ > render no crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Calendar
    currentMonth={2}
    dateStatistics={
      Object {
        "preferDate": Object {},
        "unavailableDate": Array [],
      }
    }
    filterByName="ALL"
    nextMonth={[MockFunction]}
    onDateClick={[MockFunction]}
    preferDate={Object {}}
    prevMonth={[MockFunction]}
    selectedDate={2}
    setFilterName={[MockFunction]}
    unavailableDate={Array []}
    userList={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <NameFilter
          filterByName="ALL"
          setFilterName={[MockFunction]}
          userList={Array []}
        />,
        <div
          className="calendar"
        >
          <CalendarHeader
            currentMonth={6}
            nextMonth={[Function]}
            prevMonth={[Function]}
          />
          <CalendarDays
            currentMonth={6}
          />
          <CalendarCells
            currentMonth={6}
            onDateClick={[Function]}
            preferDate={Object {}}
            selectedDate={6}
            unavailableDate={Array []}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "filterByName": "ALL",
          "setFilterName": [MockFunction],
          "userList": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <CalendarHeader
              currentMonth={6}
              nextMonth={[Function]}
              prevMonth={[Function]}
            />,
            <CalendarDays
              currentMonth={6}
            />,
            <CalendarCells
              currentMonth={6}
              onDateClick={[Function]}
              preferDate={Object {}}
              selectedDate={6}
              unavailableDate={Array []}
            />,
          ],
          "className": "calendar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "currentMonth": 6,
              "nextMonth": [Function],
              "prevMonth": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "currentMonth": 6,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "currentMonth": 6,
              "onDateClick": [Function],
              "preferDate": Object {},
              "selectedDate": 6,
              "unavailableDate": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <NameFilter
            filterByName="ALL"
            setFilterName={[MockFunction]}
            userList={Array []}
          />,
          <div
            className="calendar"
          >
            <CalendarHeader
              currentMonth={6}
              nextMonth={[Function]}
              prevMonth={[Function]}
            />
            <CalendarDays
              currentMonth={6}
            />
            <CalendarCells
              currentMonth={6}
              onDateClick={[Function]}
              preferDate={Object {}}
              selectedDate={6}
              unavailableDate={Array []}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "filterByName": "ALL",
            "setFilterName": [MockFunction],
            "userList": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <CalendarHeader
                currentMonth={6}
                nextMonth={[Function]}
                prevMonth={[Function]}
              />,
              <CalendarDays
                currentMonth={6}
              />,
              <CalendarCells
                currentMonth={6}
                onDateClick={[Function]}
                preferDate={Object {}}
                selectedDate={6}
                unavailableDate={Array []}
              />,
            ],
            "className": "calendar",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentMonth": 6,
                "nextMonth": [Function],
                "prevMonth": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentMonth": 6,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentMonth": 6,
                "onDateClick": [Function],
                "preferDate": Object {},
                "selectedDate": 6,
                "unavailableDate": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
